libraries/libfc/src/crypto/bls_utils.cpp

```diff
@@ -0,0 +1,66 @@
+#include <fc/crypto/bls_utils.hpp>
+#include <fc/crypto/common.hpp>
+#include <fc/exception/exception.hpp>
+#include <bls.hpp>
+
+namespace fc { namespace crypto { namespace blslib { namespace bls_utils {
+
+   using namespace bls;
+
+   bls_private_key generate() {
+
+     char* r = (char*) malloc(32);
+
+     rand_bytes(r, 32);
+
+     std::vector<uint8_t> v(r, r+32);
+
+     return bls_private_key(v);
+
+   }
+
+   bool verify( const bls_public_key &pubkey,
+                  const std::vector<uint8_t> &message,
+                  const bls_signature &signature){
+
+      return bls::PopSchemeMPL().Verify(pubkey._pkey, message, signature._sig);

```

Comment:
A big mistake/oversight we made in 1.0 with `recover_key` host function was that given bad input (bad signature) it would throw an exception which fails the transaction: it isn't possible for the contract to handle that failure somehow. Devs have asked for that to be fixed for years and it still isn't (entirely).

It appears this implementation has the same kind of flaw. This will at least do a `G1Eelement::FromBytes(pubkey)` and `G2Element::FromBytes(signature)`. Both of those will throw an exception on various conditions of the input. I didn't look deeper in to the actual verification.

Admittedly, I'm not familiar enough with the kind of use cases for this host function to know if there are use cases for it to be "non throwing"

Reply:
@arhag suggested we broaden the scope of this feature to expose lower level cryptographic primitives (g1/g2 add, multiply, exponentiation and pairing functions). I believe this would offer a satisfactory solution to this issue, since it would allow the implementation of custom verification or aggregation functions.

Reply:
yep I was going to mention that -- it seems like we'd want the exposed operations to be as granular as the proposed EIP-2537 so that Antelope's EVM can easily make use of them too